### PR TITLE
feat(meta): auto-clean expired keys

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

##### feat(meta): auto-clean expired keys

- Clean up to 32 expired keys before applying a raft-log.

- When modifying(insert/update/delete) a kv record, it emits at most two events:
  - If the existent key expires, emit a `delete` event because an
    expired key is seen and removed.
  - Then, the event of the update follows.

- Fix: #8489
- Fix: #8490
- Fix: #8540

## Related Issues

### CODE REVIEW

File: src/meta/raft-store/src/state_machine/sm.rs [hunk 1]
```diff
@@ -809,43 +900,109 @@ impl StateMachine {
...
+        }
+
+        // No change, no need to update expiration index
+        if prev == res {
+            return Ok((expired, prev, res));
+        }
+
+        // Remove previous expiration index, add a new one.
+
+        if let Some(sv) = &prev {
```
Comment (Xuanwo) [inline src/meta/raft-store/src/state_machine/sm.rs:942]: Let's play with `#![feature(let_chains)]`!

```
if let Some(sv) = &prev && let Some(m) = &sv.meta && let Some(exp) = m.expire_at {
    expires.remove(&ExpireKey::new(exp * 1000, sv.seq))?;
}
```

Comment (drmingdrmer) [reply src/meta/raft-store/src/state_machine/sm.rs:942]: Smart guy!

Comment (drmingdrmer) [reply src/meta/raft-store/src/state_machine/sm.rs:942]: Hmm... Clion does not recognize such syntax and refuses to format codes for me.

I'd have to leave it as is.

Comment (Xuanwo) [reply src/meta/raft-store/src/state_machine/sm.rs:942]: Tracked at https://github.com/intellij-rust/intellij-rust/issues/8625

Comment (drmingdrmer) [reply src/meta/raft-store/src/state_machine/sm.rs:942]: image

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**